### PR TITLE
bpo-43224: Implement pickling of TypeVarTuples

### PR DESCRIPTION
**Pull request by mrahtz**
Guess who learned a whole bunch about how pickling works :)

Notes:
* I was a bit horrified to realise that the class I'd renamed to `_BoundVarianceMixin` also contains a key ingredient for pickling of things that use it! I've refactored its `__reduce__` into a separate mixin for clarity.
* I'm not _entirely_ sure whether these tests make sense - is `assertEqual` a reasonable way of testing this functionality? Am I correct in thinking we do need to test both `assertEqual` and `assertIs` for TypeVarTuples and unpacked TypeVarTuples?

This PR doesn't include implementation of pickling support for unpacked native `tuple`; I'll do that in a future PR. (And I guess we'll also need to add tests for `copy`?)

https://bugs.python.org/issue43224

### CODE REVIEW

File: Lib/test/test_typing.py [hunk 1]
```diff
@@ -838,6 +838,143 @@ class C(Generic[Unpack[Ts]]): pass
         self.assertNotEqual(C[Unpack[Ts1]], C[Unpack[Ts2]])
 
 
+class TypeVarTuplePicklingTests(BaseTestCase):
```
Comment (Fidget-Spinner) [inline Lib/test/test_typing.py:926]: Holy moly these are really detailed and comprehensive tests. I was expecting something more barebones.

You might want to also take a look at how other types test pickle. They need to cover all the pickle protocol versions.
https://github.com/python/cpython/blob/5bdeab67a609308b59f53e7b5f738269eaf28095/Lib/test/test_typing.py#L5285-L5296

**Review by Fidget-Spinner — COMMENTED** [review]
> correct in thinking we do need to test both assertEqual and assertIs for TypeVarTuples and unpacked TypeVarTuples?

IIRC, `assertIs` for pickle/unpickling isn't guaranteed, so we don't need to test for that.

Other than the tests, the code in typing.py LGTM.

File: Lib/test/test_typing.py [hunk 2]
```diff
@@ -838,6 +838,143 @@ class C(Generic[Unpack[Ts]]): pass
...
+    # These are slightly awkward tests to run, because TypeVarTuples (and things
+    # made generic using TypeVarTuples) are only picklable if defined in the
+    # global scope. We therefore need to push various things defined in these
+    # tests into the global scope with `global` statements at the start of each
+    # test.
+
+    def test_typevartuple_can_be_pickled_without_error(self):
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: This test seems redundant with the next one

File: Lib/test/test_typing.py [hunk 3]
```diff
@@ -838,6 +838,143 @@ class C(Generic[Unpack[Ts]]): pass
...
+        Ts2 = pickle.loads(pickle.dumps(Ts1))
+        self.assertEqual(Ts1, Ts2)
+
+    def test_typevartuple_is_immutable_by_pickle(self):
+        global Ts1  # See explanation at start of class.
+        Ts1 = TypeVarTuple('Ts1')
+        Ts2 = pickle.loads(pickle.dumps(Ts1))
+        self.assertIs(Ts1, Ts2)
+
+    def test_unpacked_typevartuple_can_be_pickled_without_error(self):
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: Same

File: Lib/test/test_typing.py [hunk 4]
```diff
@@ -838,6 +838,143 @@ class C(Generic[Unpack[Ts]]): pass
...
+        pickle.dumps(Unpack[Ts])
+
+    def test_pickling_then_unpickling_typevartuple_results_in_equality(self):
+        global Ts  # See explanation at start of class.
+        Ts = TypeVarTuple('Ts')
+        unpacked1 = Unpack[Ts]
+        unpacked2 = pickle.loads(pickle.dumps(unpacked1))
+        self.assertEqual(unpacked1, unpacked2)
+
+    def test_variadic_class_can_be_pickled_without_error(self):
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: And again

File: Lib/test/test_typing.py [hunk 5]
```diff
@@ -838,6 +838,143 @@ class C(Generic[Unpack[Ts]]): pass
...
+
+        class A1(Generic[Unpack[Ts]]): pass
+        A2 = pickle.loads(pickle.dumps(A1))
+        self.assertEqual(A1, A2)
+
+        class B1(Generic[T, Unpack[Ts]]): pass
+        B2 = pickle.loads(pickle.dumps(B1))
+        self.assertEqual(B1, B2)
+
+    def test_tuple_with_typevartuple_can_be_pickled_without_error(self):
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: And again

**Review by JelleZijlstra — COMMENTED** [review]
Thanks, I think these tests are a bit repetitive.

File: Lib/test/test_typing.py (hunk 1 above)
Comment (mrahtz) [reply Lib/test/test_typing.py:926]: Ah, good point. I felt a bit uneasy about having loops in all these tests, so I've implemented it using a decorator. (I've also gone through and changed the other tests cases to use the decorator too, making the test bodies a bit simpler, but let me know if you'd prefer this done in a separate PR/not at all, and I'll revert that commit.)

File: Lib/test/test_typing.py (hunk 2 above)
Comment (mrahtz) [reply Lib/test/test_typing.py]: Fair, I was feeling a bit uncertain about this myself - recently I've been erring on the side of trying to separate out tests for different behaviour as much as possible, but happy to get rid of these if you're think they're more trouble than they're worth :)

File: Lib/test/test_typing.py (hunk 4 above)
Comment (mrahtz) [reply Lib/test/test_typing.py]: Done.

File: Lib/test/test_typing.py (hunk 3 above)
Comment (mrahtz) [reply Lib/test/test_typing.py]: Done.

File: Lib/test/test_typing.py (hunk 5 above)
Comment (mrahtz) [reply Lib/test/test_typing.py]: Done.

File: Lib/test/test_typing.py (hunk 1 above)
Comment (Fidget-Spinner) [reply Lib/test/test_typing.py:926]: Thanks. The refactoring is good. But yeah I'd prefer them done in another PR please. There're some points of contention.

Comment (mrahtz) [reply Lib/test/test_typing.py:926]: Done.

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. Thanks!

**Comment by Fidget-Spinner** [comment]
Why was the PR closed @mrahtz ?

**Comment by mrahtz** [comment]
Lol, my bad - I was cleaning up some old branches, and forgotten this hadn't yet been merged 😅

File: Lib/typing.py [hunk 6]
```diff
@@ -987,11 +992,18 @@ class C(Generic[*Ts]): ...
       C[()]        # Even this is fine
 
     For more details, see PEP 646.
+
+    Note that only type variables tuples defined in global scope can be pickled.
```
Comment (JelleZijlstra) [inline Lib/typing.py]:
```suggestion
    Note that only TypeVarTuples defined in global scope can be pickled.
```

File: Lib/test/test_typing.py [hunk 7]
```diff
@@ -894,6 +906,93 @@ class C(Generic[Unpack[Ts]]): pass
...
+        Ts = TypeVarTuple('Ts')
+        unpacked1 = Unpack[Ts]
+        unpacked2 = pickle.loads(pickle.dumps(unpacked1, proto))
+        self.assertEqual(unpacked1, unpacked2)
+
+    @all_pickle_protocols
+    def test_pickling_then_unpickling_variadic_class_results_in_equality(
+            self, proto
+    ):
+        global T, Ts, A1, B1  # See explanation at start of class.
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: This misses the point a bit. Classes are always pickled by name, so it doesn't really matter what Ts is doing here. I think we can simply remove this test.

File: Lib/test/test_typing.py [hunk 8]
```diff
@@ -894,6 +906,93 @@ class C(Generic[Unpack[Ts]]): pass
         self.assertNotEqual(C[Unpack[Ts1]], C[Unpack[Ts2]])
 
 
+class TypeVarTuplePicklingTests(BaseTestCase):
+    # These are slightly awkward tests to run, because TypeVarTuples (and things
+    # made generic using TypeVarTuples) are only picklable if defined in the
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: The parenthetical isn't really accurate; I suggest removing it.

File: Lib/test/test_typing.py [hunk 9]
```diff
@@ -894,6 +906,93 @@ class C(Generic[Unpack[Ts]]): pass
...
+    # global scope. We therefore need to push various things defined in these
+    # tests into the global scope with `global` statements at the start of each
+    # test.
+
+    @all_pickle_protocols
+    def test_pickling_then_unpickling_results_in_equality(self, proto):
+        global Ts1  # See explanation at start of class.
+        Ts1 = TypeVarTuple('Ts1')
+        Ts2 = pickle.loads(pickle.dumps(Ts1, proto))
+        self.assertEqual(Ts1, Ts2)
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: We should use `assertIs` instead to assert we get the exact same TVT back.

File: Lib/test/test_typing.py [hunk 10]
```diff
@@ -894,6 +906,93 @@ class C(Generic[Unpack[Ts]]): pass
...
+
+    @all_pickle_protocols
+    def test_pickling_then_unpickling_variadic_func_results_in_equality(
+            self, proto
+    ):
+        global T, Ts, a1, b1, c1, d1, e1  # See explanation at start of class.
+        T = TypeVar('T')
+        Ts = TypeVarTuple('Ts')
+
+        def a1(*args: Unpack[Ts]): pass
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: As with classes, functions are pickled by name, so this test doesn't really test anything about TypeVarTuples.

File: Lib/test/test_typing.py (hunk 7 above)
Comment (mrahtz) [reply Lib/test/test_typing.py]: Done.

File: Lib/test/test_typing.py (hunk 8 above)
Comment (mrahtz) [reply Lib/test/test_typing.py]: Done.

File: Lib/test/test_typing.py (hunk 9 above)
Comment (mrahtz) [reply Lib/test/test_typing.py]: Done. (I'm assuming `assertEqual` is still correct for the tuples, but re-open if I'm wrong.)

File: Lib/test/test_typing.py (hunk 10 above)
Comment (mrahtz) [reply Lib/test/test_typing.py]: Done.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by mrahtz** [comment]
Thanks, Jelle!

**Pull request merged**